bftengine/src/bftengine/PersistentStorageImp.hpp

```diff
@@ -0,0 +1,220 @@
+// Concord
+//
+// Copyright (c) 2019 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of sub-components with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the sub-component's license, as noted in the LICENSE file.
+
+#pragma once
+
+#include "PersistentStorage.hpp"
+#include "MetadataStorage.hpp"
+#include "ReplicaConfigSerializer.hpp"
+#include "PersistentStorageWindows.hpp"
+
+namespace bftEngine {
+namespace impl {
+
+enum ConstMetadataParameterIds {
+  VERSION_PARAMETER = 0,
+  FETCHING_STATE,
+  LAST_EXEC_SEQ_NUM,
+  PRIMARY_LAST_USED_SEQ_NUM,
+  LOWER_BOUND_OF_SEQ_NUM,
+  LAST_VIEW_TRANSFERRED_SEQ_NUM,
+  LAST_STABLE_SEQ_NUM,
+  REPLICA_CONFIG,
+  CONST_METADATA_PARAMETERS_NUM
+};
+
+const uint16_t seqWinSize = kWorkWindowSize;
+
+constexpr uint16_t numOfSeqNumWinParameters = SeqNumData::getNumOfParams();
+const uint16_t numOfSeqNumWinObjs = seqWinSize * numOfSeqNumWinParameters + 1;
+
+const uint16_t checkWinSize = (kWorkWindowSize + checkpointWindowSize) / checkpointWindowSize;
+
+constexpr uint16_t numOfCheckWinParameters = CheckData::getNumOfParams();
+const uint16_t numOfCheckWinObjs = checkWinSize * numOfCheckWinParameters + 1;
+
+enum WinMetadataParameterIds {
+  SEQ_NUM_WINDOW = CONST_METADATA_PARAMETERS_NUM,
+  CHECK_WINDOW = SEQ_NUM_WINDOW + numOfSeqNumWinObjs,
+  WIN_PARAMETERS_NUM = CHECK_WINDOW + numOfCheckWinObjs
+};
+
+// LAST_EXIT_FROM_VIEW_DESC contains up to kWorkWindowSize descriptor objects
+// (one per PrevViewInfo) plus one - for simple descriptor parameters.
+const uint16_t numOfLastExitFromViewDescObjs = kWorkWindowSize + 1;
+
+// LAST_NEW_VIEW_DESC contains numOfReplicas_ (2 * f + 2 * c + 1) descriptor
+// objects plus one - for simple descriptor parameters.
+enum DescMetadataParameterIds {
+  LAST_EXIT_FROM_VIEW_DESC = WIN_PARAMETERS_NUM,
+  LAST_EXEC_DESC = LAST_EXIT_FROM_VIEW_DESC + numOfLastExitFromViewDescObjs,
+  LAST_NEW_VIEW_DESC
+};
+
+class PersistentStorageImp : public PersistentStorage {
```

Comment:
It has a destructor. This class is a kind of singleton. It is not expected to be copied or assigned, but I agree that it would be safer to make them deleted.

Reply:
I figured as much. I just worry about implicit copy by accident. It probably is not an issue here as you say, but since Alex pointed this out on some of the closed source code it's been in my mind to be on the lookout.